text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+resources that are part of the AWS CDK framework.
+
+There are more dependencies like these that need to be added to the framework,
+but we are currently blocked on adding them with the current design because they
+would increase the size too much. The current known list is:
+
+1. Additional versions of
+   [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl). This is a
+   pressing problem for customers using the EKS module, who are limited to a
+   single version of the kubectl CLI today.
+2. Additional versions of [awscli](https://pypi.org/project/awscli/). See
+   [issue](https://github.com/aws/aws-cdk/issues/13993).
+3. [Boto3](https://pypi.org/project/boto3/), and
+   [aws-sdk](https://www.npmjs.com/package/aws-sdk). All custom resources in the
+   AWS CDK framework currently rely on whatever version of the AWS SDK that is
+   available in the Lambda runtime by default. To make the custom resources more
+   robust, and allow us to choose the exact version of the AWS SDK used by them,
+   we should bundle these dependencies into each custom resource’s Lambda code.
+
+If we add all of these with the current design, that would increase the size of
+`aws-cdk-lib` by approximately 130 MiB*, and make the Lambda Layers the biggest
+contributor of size at 39%.
+
+*See Appendix A to see where this estimate came from.
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| Lambda Layer zip files         | 38.83%                         |
+| .jsii files                    | 30.14%                         |
+| Source Maps (.js.map)          | 15.87%                         |
+| Javascript code (.js)          | 6.63%                          |
+| Type Declaration files (.d.ts) | 6.62%                          |
+| README, etc (.md)              | 0.55%                          |
+| bundled npm dependencies       | 0.50%                          |
+| .ts-fixture                    | 0.27%                          |
+| .jsiirc                        | 0.26%                          |
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
+
+### What is the technical solution (design) of this feature?
+
+There are different categories of files contributing to the size of aws-cdk-lib,
+and each category will have it’s own solution. The following sections cover the
+top five categories in size contribution in order from largest to smallest
+potential size.
+
+#### Lambda Layer Zip Files
+
+Use separate npm packages for each `lambda-layer-X` module, which the CDK CLI
+will automatically install if not already present in the `node_modules`
```

Comment:
Needs a bit of elaboration on how the CLI will know what it needs to install, and how the app will know what has been installed and where.

Reply:
I am no longer planning to use the CLI for installing/downloading.